pkg/uroot/bb.go

```diff
@@ -223,109 +209,133 @@ func BBBuild(opts BuildOpts) (ArchiveFiles, error) {
 }
 
 type CommandTemplate struct {
-	Gopath  string
 	CmdName string
-	Init    string
 }
 
 type Package struct {
 	name string
-	pkg  *build.Package
-	fset *token.FileSet
-	ast  *ast.Package
 
+	pkg      *build.Package
+	fset     *token.FileSet
+	ast      *ast.Package
+	typeInfo types.Info
+	types    *types.Package
+
+	// initCount keeps track of what the next init's index should be.
 	initCount uint
-	init      string
+
+	// init is the cmd.Init function that calls all other InitXs in the
+	// right order.
+	init *ast.FuncDecl
+
+	// initAssigns is a map of assignment expression -> assignment
+	// statement.
+	//
+	// types.Info.InitOrder keeps track of Initializations by Lhs name and
+	// Rhs ast.Expr.  We reparent the Rhs in assignment statements in InitX
+	// functions, so we use the Rhs as an easy key here.
+	// types.Info.InitOrder + initAssigns can then easily be used to derive
+	// the order of AssignStmts.
+	//
+	// The key Expr must also be the AssignStmt.Rhs[0].
+	initAssigns map[ast.Expr]*ast.AssignStmt
+}
+
+func (p *Package) nextInit() *ast.Ident {
+	i := ast.NewIdent(fmt.Sprintf("Init%d", p.initCount))
+	p.init.Body.List = append(p.init.Body.List, &ast.ExprStmt{X: &ast.CallExpr{Fun: i}})
+	p.initCount++
+	return i
 }
 
 func (p *Package) CommandTemplate() CommandTemplate {
 	return CommandTemplate{
-		Gopath:  p.pkg.Root,
 		CmdName: p.name,
-		Init:    p.init,
 	}
 }
 
-func (p *Package) rewriteFile(opts BuildOpts, f *ast.File) bool {
-	// Inspect the AST and change all instances of main()
-	var pos token.Position
+// TODO:
+// - write an init name generator, in case InitN is already taken.
```

Comment:
I like how Grumpy prefixes generated identifiers with a Greek symbol to indicate they've been generated and not a character a human would type. For example, ßInit. Something like that could be used here.

Reply:
It's not a general solution (people can name Go functions anything they want), so I'd rather find some other way. But maybe the default can be something like this.